### PR TITLE
Rover: Add a line-following mode [WIP]

### PR DESCRIPTION
**Pull request by rishabsingh3003**
This PR is in response to #7590.

- I have attempted to simulate Analog IR Sensor's [https://www.sparkfun.com/products/9453]. These sensors are widely used in line following robots - white surfaces reflect much more light than black, so, when directed towards a white surface, the voltage output will be lower than that on a black surface. A minimum of two of these sensors are required on the far edges of the robot to determine where the vehicle is with respect to the black line.  
image

- A new IR Sensor library is created which is divided into front-end, back-end, and an analog driver similar to other standard Ardupilot sensor libraries. This will make it easier to add other line following sensors to the library. Certain parameters are also defined which need to be specified by the user. Using the values that are picked up from the sensors, an error is generated to approximate the position of the vehicle with respect to the line. **The first 8 commits include these additions**

- A new "Line Follower" mode is created in the **last 10 commits**. The IR sensors are read periodically and the generated error is directly fed to ` AR_AttitudeControl::get_steering_out_rate()`.
The following image explains the theory behind how this mode works. 

image

- **TESTING:**

1. The simulation in SITL works well. To simulate, first change the param's :
 `param set IR_PIN_1 16`
`param set IR_PIN_2 17`
`param set IR_ADC_VOLTAGE 5`

Then write the commands: 
`mode 17`
`arm throttle`
`rc 3 1700`

The GIF of a simulation is posted below:
Line_Follower_SITL

2. `Tools/scripts/build_all.sh` script was run. All systems were built successfully. 

3. I had attached the analog sensors to the 3.3V ADC of my Pixhawk, The values were read correctly. 

4. Unfortunately, I do not have a compatible motor driver (ESC) to build the rover, therefore I **can not test the mode on an actual vehicle** but this PR should at least give a good start to this age-long issue. I had tested a similar algorithm on an Arduino Uno based vehicle with a cheap motor driver and the mode worked well.

- @peterbarker please tell me the modifications required in this PR, if they are needed. I realize there are a lot of commits! Hopefully, this will be a start to Ardupilot's Line Follower vehicles and other developers can contribute to this.

### CODE REVIEW

File: libraries/AP_HAL_SITL/sitl_irsensor.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    bool projected_line_vector = projected_home_line.get_vector_xy_from_origin_NE(end_NE);    
+    bool sensor_1_vector = sensor_1.get_vector_xy_from_origin_NE(sensor_1_offset_NE);
+    bool sensor_2_vector = sensor_2.get_vector_xy_from_origin_NE(sensor_2_offset_NE); 
+
+    float crosstrack_error_1 = Vector2f::closest_distance_between_line_and_point(start_NE, end_NE, sensor_1_offset_NE); // distance between individual sensors and the virtual line
+    float crosstrack_error_2 = Vector2f::closest_distance_between_line_and_point(start_NE, end_NE, sensor_2_offset_NE);
+
+    if (crosstrack_error_1>= max_sensor_range){ //force sensor to stay in range limit
+        crosstrack_error_1 = max_sensor_range;
+    }if(crosstrack_error_2>= max_sensor_range){
```
Comment (khancyr) [inline libraries/AP_HAL_SITL/sitl_irsensor.cpp]: Syntax issue here

File: libraries/AP_HAL_SITL/sitl_irsensor.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
+/*
+  SITL handling
+
+  This simulates a Analog IR Sensor
+
+ */
+
```
Comment (khancyr) [inline libraries/AP_HAL_SITL/sitl_irsensor.cpp]: This file need a reformat asp plenties space and float qualifier are missing

File: libraries/AR_IR/AR_IR_Sensor.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+void AR_IR::init(){
+	drivers = new AR_IR_Analog(*this, state);
+}
+
+
+void AR_IR::read(){
+	drivers->read();
+}
+
+void AR_IR:: calc_line_error(float &line_error){
```
Comment (khancyr) [inline libraries/AR_IR/AR_IR_Sensor.cpp]: The file also need a reformat

File: libraries/AR_IR/AR_IR_Sensor.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#define IR_1_PIN_NUM  16
```
Comment (khancyr) [inline libraries/AR_IR/AR_IR_Sensor.h]: Those don't sems to be used

Comment (khancyr) [reply libraries/AR_IR/AR_IR_Sensor.h]: Ha I saw the usage...

**Comment by rishabsingh3003** [comment]
@khancyr Thanks for the quick review!
As I understand your suggestions, you want me to make a backend for SITL instead of changing HAL files itself. That makes a lot of sense. I'll fix the formatting issues too. Will take some time, though.

**Comment by khancyr** [comment]
Yes, I will confirm with @peterbarker as I think it will be simplier

**Comment by rishabsingh3003** [comment]
@khancyr I dropped all the HAL_SITL commits and moved them to IR backend instead as you asked (Therefore HAL is untouched now). I have formatted the new files as well. Hope they are okay this time. Do let me know if any changes are required.

**Comment by rishabsingh3003** [comment]
This probably belongs as a Lua script now! The implemented algorithm seemed okay, so if anyone wants to pick on this in the future, please go ahead! I am closing this PR.

**Pull request closed without merging**